lib/model-builder.js

```diff
@@ -656,41 +656,27 @@ ModelBuilder.prototype.define = function defineClass(className, properties, sett
 };
 
 function createModelClassCtor(name, ModelBaseClass) {
-  // A simple sanitization to handle most common characters
-  // that are used in model names but cannot be used as a function/class name.
-  // Note that the rules for valid JS indentifiers are way too complex,
-  // implementing a fully spec-compliant sanitization is not worth the effort.
-  // See https://mathiasbynens.be/notes/javascript-identifiers-es6
+  // Backwards-compatible replacements of common characters.
+  // https://github.com/strongloop/loopback-datasource-juggler/pull/1534
   name = name.replace(/[-.:]/g, '_');
 
-  try {
-    // It is not possible to access closure variables like "ModelBaseClass"
-    // from a dynamically defined function. The solution is to
-    // create a dynamically defined factory function that accepts
-    // closure variables as arguments.
-    const factory = new Function('ModelBaseClass', `
-      // every class can receive hash of data as optional param
-      return function ${name}(data, options) {
-        if (!(this instanceof ${name})) {
-          return new ${name}(data, options);
-        }
-        if (${name}.settings.unresolved) {
-          throw new Error(g.f('Model %s is not defined.', ${JSON.stringify(name)}));
-        }
-        ModelBaseClass.apply(this, arguments);
-      };`);
-
-    return factory(ModelBaseClass);
-  } catch (err) {
-    // modelName is not a valid function/class name, e.g. 'grand-child'
-    // and our simple sanitization was not good enough.
-    // Falling back to legacy 'ModelConstructor' name.
-    if (err.name === 'SyntaxError') {
-      return createModelClassCtor('ModelConstructor', ModelBaseClass);
-    } else {
-      throw err;
+  // every class can receive hash of data as optional param
+  function constructor(data, options) {
+    if (!(this instanceof constructor)) {
+      return new constructor(data, options);
+    }
+    if (constructor.settings.unresolved) {
+      throw new Error(g.f('Model %s is not defined.', name));
     }
+    ModelBaseClass.apply(this, arguments);
   }
+
+  Object.defineProperty(constructor, 'name', {
```

Comment:
I am surprised this works. Can you please verify that the `name` property is used by error stack traces too? In the past, it was not possible to override function name this way.

Reply:
It's also to verify this works on all Node.js versions we support, I believe that means Node.js 8.x, 10.x and 11.x for the `master` branch.

Reply:
> Can you please verify that the `name` property is used by error stack traces too?

It is.

> It's also important to verify this works on all Node.js versions we support, I believe that means Node.js 8.x, 10.x and 11.x for the master branch.

It has worked since Node 4.